geps/gep-1731/index.md

```diff
@@ -0,0 +1,409 @@
+
+    * `retriable-status-codes` Allows specifying any response code.
+
+    * `retriable-headers` Allows specifying headers in the response for which a request should be retried.
+
+    * `http3-post-connect-failure` Retry if a request is sent over HTTP/3 to the upstream server and fails after connecting.
+
+  * [**gRPC**](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/router_filter#x-envoy-retry-grpc-on)
+
+    * `cancelled`
+    * `deadline-exceeded`
+    * `internal`
+    * `resource-exhausted`
+    * `unavailable`
+
+* `num_retries` The allowed number of retries, defaults to 1. Further notes on specific behavior can be found under [`x-envoy-max-retries`](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/router_filter#x-envoy-max-retries).
+
+* `per_try_timeout` Specifies a non-zero upstream timeout per retry attempt (including the initial attempt). If left unspecified, Envoy will use the global route timeout for the request. Equivalent to the Gateway API HTTPRouteRule BackendRequest timeout.
+
+* `per_try_idle_timeout` This timeout is useful in cases in which total request timeout is bounded by a number of retries and a per_try_timeout, but there is a desire to ensure each try is making incremental progress.
+
+* ...several options for retry priority and host selection, likely not portable
+
+* `retriable_status_codes` Allows specifying any response code for which a request should be retried.
+
+* `retry_back_off` Allows specifying a default base interval and max interval for retry attempts.
+
+* `rate_limited_retry_back_off` Advanced configuration for adjusting retry attempt timing based on headers sent in the upstream server response. Alternative to the default exponential back off strategy.
+
+* `retriable_headers` HTTP response headers for which a backend request should be retried.
+
+* `retriable_request_headers` HTTP headers which must be present in the _request_ for retries to be attempted.
+
+Supports configuring a [RetryBudget](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/circuit_breaker.proto#envoy-v3-api-msg-config-cluster-v3-circuitbreakers-thresholds-retrybudget) with a following parameters in cluster CircuitBreaker thresholds.
+
+* `budget_percent` Specifies the limit on concurrent retries as a percentage of the sum of active requests and active pending requests. For example, if there are 100 active requests and the budget_percent is set to 25, there may be 25 active retries. This parameter is optional. Defaults to 20%.
+
+* `min_retry_concurrency` Specifies the minimum retry concurrency allowed for the retry budget. The limit on the number of active retries may never go below this number. This parameter is optional. Defaults to 3.
+
+#### NGINX
+
+The [`proxy_next_upstream`](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_next_upstream) directive specifies in which cases a request should be passed to the next server:
```

Comment:
Something important to note with this nginx directive is that it will not retry the current server. So if your backend application has 1 replica, if it fails, there are no retries. If your backend applications has multiple replicas, nginx will simply retry by passing the request to each backend server until there are no more or a successful response is returned.

To get real retry logic in nginx as desired in this GEP, a custom module (NJS or otherwise) will likely need to be built.